### ISSUE TITLE
issue: Errors are not removed from array values

### ISSUE THREAD

**Issue by marcospassos** [react-hook-form/react-hook-form#9867]
### Version Number

7.43.0/8-next.4

### Codesandbox/Expo snack

https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-w9kye1?file=/src/index.js

### Steps to reproduce

1. Click on "Add friend" 3 times and see an error is shown
2. Click on "Remove friend" and see the error is still shown

### Expected behaviour

Revalidate the form once the item is removed to maintain consistency with the `append` behavior.

Notice that using `trigger` is not an option since it will trigger the validation of all nested values, which is undesired. In our case, we've deeply nested forms, and doing that results in a terrible UX.

https://user-images.githubusercontent.com/943036/216787017-1fa5f3a5-a766-4c91-8e73-5d73076e5a76.mov

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Please use the following code to debug the issue first:

```
resolver: async (data, context, options) => {
  // you can debug your validation schema here
  console.log('formData', data)
  console.log('validation result', await anyResolver(schema)(data, context, options))
  return anyResolver(schema)(data, context, options)
},
```

https://react-hook-form.com/api/useform

![Screenshot 2023-02-05 at 9 46 01 am](https://user-images.githubusercontent.com/10513364/216792644-ae1c1876-948c-4b15-87a5-2fd73f83aaa8.png)

**Comment by marcospassos** [comment]
I've updated the example with this debugging part. As you can see below, the schema correctly reports no error when removing an item, but RHF still shows the error:

![image](https://user-images.githubusercontent.com/943036/216792873-b385bf32-65b4-42bf-a51e-77b834461af6.png)

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

Please take a look at the CSB, I have switched to the latest version and let me know if I am missing something

Uploading Screen Recording 2023-02-05 at 10.09.19 am.mov…

**Comment by bluebill1049** [comment]
related issue: https://github.com/react-hook-form/react-hook-form/issues/9868

**Comment by marcospassos** [comment]
The problem is still present in your CSB fork:

https://user-images.githubusercontent.com/943036/216824020-04c00127-d06f-4759-a606-2ee5edf2bb68.mov

Notice that the error correctly appears after adding the third element (the max is 2), but after removing it, the error remains.

**Comment by marcospassos** [comment]
> related issue: #9868

Also, not sure these problems are related.

**Comment by bluebill1049** [comment]
<img width="579" alt="Screenshot 2023-02-06 at 7 12 29 am" src="https://user-images.githubusercontent.com/10513364/216842447-974b7781-682a-4e3a-b607-c0e2229f67e1.png">

Your schema is still returning errors even with 2 items.

**Comment by marcospassos** [comment]
It's returning the error for the right fields: the array elements. There's no error for the array itself, but the RHF still shows an error at the array level:

![image](https://user-images.githubusercontent.com/943036/216847005-6a963d0b-f0b2-49cd-8d2c-453a61296bf3.png)

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/issues/9868#issuecomment-1418270268

**Comment by marcospassos** [comment]
Sorry, I still don't get it.

Are you saying that RHF showing an error that no longer exists is right? As I have shown, the last revalidation resulted in no error at the array level, but it still keeps a state with an error from the past. 

Even using the `dirtyFields` or `thouchedFields`, I can't workaround this issue since the fields are touched and dirty, but the error doesn't exist - the state that's outdated.

**Comment by bluebill1049** [comment]
> Are you saying that RHF showing an error that no longer exists is right? As I have shown, the last revalidation resulted in no error at the array level, but it still keeps a state with an error from the past.

No, The error is still there under the field array key `firends`.

**Comment by marcospassos** [comment]
No, there is no error under the field array key, please see:

https://user-images.githubusercontent.com/943036/216847735-2bdd4d0b-85f9-4128-adc2-aa220d71a69a.mov

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

https://user-images.githubusercontent.com/10513364/216848063-40929aeb-d1ac-4752-b9bb-8013eb7f4ed1.mov

- 3-row trigger error
- remove 2 still have errors from the schema
- error is hence not removed

**Comment by marcospassos** [comment]
So, you're saying that an error from a parent field will only be removed once all nested field errors have been fixed, even though it no longer has errors?

**Comment by bluebill1049** [comment]
https://github.com/react-hook-form/react-hook-form/issues/9868#issuecomment-1418270268

> pick up the errors by field array key

**Comment by marcospassos** [comment]
Sorry, I didn't get what you mean by "pick up the errors by field array key".

Do I have access to the path of the field that triggered the error from the resolver?

**Comment by bluebill1049** [comment]
> Do I have access to the path of the field that triggered the revalidation from the resolver?

I think you could try that which is what i commented in the above rely as well.

> revalidate the schema (you may be able to do a custom resolver here, to resolve your error, so meet your UX logic)

**Comment by bluebill1049** [comment]
I think we do return the validation field path. Sorry on my phone now reply would be limited.

**Comment by marcospassos** [comment]
No matter what I do, the error doesn't disappear. Even not returning any other error, RHF still shows the error:
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-w9kye1?file=/src/index.js

https://user-images.githubusercontent.com/943036/216850373-cc922676-ee94-4a0b-b256-2f9d89bd9a33.mov

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/react-hook-form-array-fields-with-validation-yup-forked-ztx800?file=/src/index.js

```tsx
import React from "react";
import ReactDOM from "react-dom/client";
import { useFieldArray, useForm } from "react-hook-form";
import * as yup from "yup";
import { yupResolver } from "@hookform/resolvers/yup";
import "./styles.css";

const schema = yup.object().shape({
  friends: yup
    .array()
    .of(
      yup.object().shape({
        firstName: yup.string().required("First Name is required")
      })
    )
    .min(1)
    .max(2)
});

function App() {
  const { register, handleSubmit, control, formState, clearErrors } = useForm({
    mode: "onChange",
    reValidateMode: "onChange",
    defaultValues: {
      friends: []
    },
    resolver: async (data, context, options) => {
      // you can debug your validation schema here
      const { errors, values } = await yupResolver(schema)(
        data,
        context,
        options
      );

      if (Array.isArray(errors.friends) && !formState.isSubmitted) {
        return {
          errors: {},
          values
        };
      }

      return { errors, values };
    }
  });

  console.log("formState.errors", formState.errors);

  const { fields, remove, append } = useFieldArray({
    name: "friends",
    control: control
  });

  const onSubmit = (data) => {
    console.log("submit", data);
  };

  const addFriend = () => {
    append({ firstName: "" });
    clearErrors("friends");
  };

  const removeFriend = (index) => () => {
    remove(index);
  };

  const clearFriends = () => {
    remove();
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <p>{formState?.errors?.friends?.message}</p>
      {fields.map((field, index) => {
        return (
          <fieldset key={field.id}>
            <label>
              First Name {index}:
              <input type="text" {...register(`friends.${index}.firstName`)} />
            </label>
            <button type="button" onClick={removeFriend(index)}>
              Remove
            </button>
          </fieldset>
        );
      })}

      <button type="button" onClick={addFriend}>
        Add Friend
      </button>
      <button type="button" onClick={clearFriends}>
        Clear Friends
      </button>
      <input type="submit" />
    </form>
  );
}

const root = ReactDOM.createRoot(document.getElementById("root"));
root.render(<App />);

```

**Comment by marcospassos** [comment]
Once you submit, the error never goes away:
https://user-images.githubusercontent.com/943036/216850778-94bfe95b-ff3b-4cfc-9c00-879709dc9c91.mov

**Comment by bluebill1049** [comment]
I am just giving an example above, it's not the full implementation detail.

**Comment by marcospassos** [comment]
Would you have any clue how I can make the error disappear?

No matter what I return from the resolver, once the error is shown, the only way to make it disappear is to trigger the whole field validation, which I want to avoid.
